dali/operators/random/noise/salt_and_pepper_noise.h

```diff
@@ -0,0 +1,154 @@
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_RANDOM_NOISE_SALT_AND_PEPPER_NOISE_H_
+#define DALI_OPERATORS_RANDOM_NOISE_SALT_AND_PEPPER_NOISE_H_
+
+#include <limits>
+#include "dali/operators/random/rng_base.h"
+#include "dali/operators/random/rng_base_cpu.h"
+#include "dali/operators/random/rng_base_gpu.h"
+#include "dali/pipeline/operator/arg_helper.h"
+
+#define DALI_SALT_AND_PEPPER_NOISE_TYPES \
+  uint8_t, int8_t, uint16_t, int16_t, uint32_t, int32_t, uint64_t, int64_t, float
+
+namespace dali {
+
+template <typename Backend, typename T>
+class SaltAndPepperNoiseImpl {
+ public:
+  using DistType = typename std::conditional_t<std::is_same<Backend, GPUBackend>::value,
+                                               curand_uniform_dist<float>,
+                                               std::uniform_real_distribution<float>>;
+  static constexpr T kDefaultSalt =
+      std::is_floating_point<T>::value ? T(1) : std::numeric_limits<T>::max();
+  static constexpr T kDefaultPepper =
+      std::is_floating_point<T>::value ? T(-1) : std::numeric_limits<T>::min();
+
+  DALI_HOST_DEV explicit SaltAndPepperNoiseImpl(float noise_prob = 0.05,
+                                                float salt_to_pepper_prob = 0.5,
+                                                T salt_val = kDefaultSalt,
+                                                T pepper_val = kDefaultPepper)
+      : noise_prob_(clamp<float>(noise_prob, 0, 1)),
+        salt_prob_(noise_prob_ * clamp<float>(salt_to_pepper_prob, 0, 1)),
+        salt_val_(salt_val), pepper_val_(pepper_val) {
+  }
+
+  template <typename Generator>
+  DALI_HOST_DEV float Generate(T input, Generator &st) {
+    (void) input;  // this noise doesn't depend on the input
+    return dist_(st);
+  }
+
+  DALI_HOST_DEV void Apply(T &output, T input, float n) {
+    if (n < noise_prob_) {
+      if (n < salt_prob_) {
+        output = salt_val_;
+      } else {
+        output = pepper_val_;
+      }
+    } else {
+      output = input;
+    }
+  }
+
+ private:
+  float noise_prob_;
+  float salt_prob_;
+  T salt_val_;
+  T pepper_val_;
+  DistType dist_;
+};
+
+template <typename Backend>
+class SaltAndPepperNoise : public RNGBase<Backend, SaltAndPepperNoise<Backend>, true> {
+ public:
+  using BaseImpl = RNGBase<Backend, SaltAndPepperNoise<Backend>, true>;
+
+  template <typename T>
+  using Impl = SaltAndPepperNoiseImpl<Backend, T>;
+
+  explicit SaltAndPepperNoise(const OpSpec &spec)
+      : BaseImpl(spec),
+        prob_("prob", spec),
+        salt_to_pepper_prob_("salt_to_pepper_prob", spec),
+        salt_val_("salt_val", spec),
+        pepper_val_("pepper_val", spec),
+        per_channel_(spec.GetArgument<bool>("per_channel")) {
+    if (prob_.IsDefined() || salt_to_pepper_prob_.IsDefined() ||
+        salt_val_.IsDefined() || pepper_val_.IsDefined()) {
+      backend_data_.ReserveDistsData(sizeof(Impl<double>) * max_batch_size_);
```

Comment:
Why we conditionally reserve space?

Reply:
Isn't it because without them we run use a default distribution?

Reply:
Yes, it is.